### PR TITLE
Improve connection error handling

### LINKED ISSUE #564: [BUG] Better error reporting when failing to authenticate via plink on Windows
## Bug summary

When plink fails to authenticate on Windows, rdiff-backup exits with a very unhelpful message.

On 2.0.5, it exited with a "Integer overflow while attempting ..." message.

On 2.1.0a1 it exits with a MemoryError in "rdiff_backup\connection.py", line 273, in _read

## Version, Python, Operating System

```yaml
exec:
  api_version:
    actual: 0
    default: 200
    max: 201
    min: 200
  argv:
  - C:\Path\To\rdiff-backup.exe
  - info
  version: 2.1.0a1
python:
  executable: C:\Path\To\rdiff-backup.exe
  name: cpython
  version: 3.9.1
system:
  fs_encoding: utf-8
  platform: Windows-10-10.0.18362-SP0
```

## rdiff-backup call

I generally call it with a schema like: `C:\Path\To\PuTTY\plink.exe %s -l Roman -C C:\Remote\rdiff-backup.exe --server` (or `server` for 2.1)

### PR DESCRIPTION
**Pull request by ericzolf**
FIX: improve handling of connection errors with clearer message, closes #564

### CODE REVIEW

File: src/rdiff_backup/connection.py [hunk 1]
```diff
@@ -288,16 +288,31 @@ def _get(self):
...
-                                      "probably originated remotely)")
+            raise ConnectionReadError(
+                "Truncated header <{hdr}> "
+                "(problem probably originated remotely)".format(
+                    hdr=header_string))
         format_string = header_string[0:1]
         req_num = self._b2i(header_string[1:2])
         length = self._b2i(header_string[2:])
 
+        if format_string not in b"qobfirRQc":
```
Comment (ikus060) [inline src/rdiff_backup/connection.py:299]: I need explanation here. why `qobfirRQc` ? Where did you get this from ?

Comment (ikus060) [reply src/rdiff_backup/connection.py:299]: Nevermind, the code was collapse. 
The following seams more explicit to me:
```

if format_string == b"q":
    raise ConnectionQuit("Received quit signal")

try:
        if format_string == b"o":
            result = pickle.loads(self._read(length))
        elif format_string == b"b":
            result = self._read(length)
        elif format_string == b"f":
            result = VirtualFile(self, self._b2i(self._read(length)))
        elif format_string == b"i":
            result = iterfile.FileToMiscIter(
                VirtualFile(self, self._b2i(self._read(length))))
        elif format_string == b"r":
            result = self._getrorpath(self._read(length))
        elif format_string == b"R":
            result = self._getrpath(self._read(length))
        elif format_string == b"Q":
            result = self._getqrpath(self._read(length))
        elif format_string == b"c":
            result = Globals.connection_dict[self._b2i(self._read(length))]
        else:
            raise ConnectionReadError(
                "Format character '{form}' invalid in <{hdr}> "
                "(problem probably originated remotely)".format(
                    hdr=header_string, form=format_string))
except MemoryError:
    raise ConnectionReadError(
        "Impossibly high data amount evaluated in header <{hdr}> "
        "(problem probably originated remotely)".format(
            hdr=header_string))
```

So whenever we are adding a new formatString, we don't need to "think" about adding it in the validation.

Comment (ericzolf) [reply src/rdiff_backup/connection.py:299]: Good point with the maintenance. I actually wanted to have the simplest check first, but it's just cosmetic, I'll change it.

Comment (ericzolf) [reply src/rdiff_backup/connection.py]: OK, addressed.
I preferred to keep the `_read` separate, IMHO it makes a more readable code, and also one easier to debug (I can check the content of data before it goes into the `_getxxx` function).

**Pull request merged**